pkg/networkserver/adr_internal_test.go

```diff
@@ -465,25 +471,22 @@ func TestAdaptDataRate(t *testing.T) {
 			},
 		},
 	} {
-		t.Run(tc.Name, func(t *testing.T) {
-			a := assertions.New(t)
-
-			dev := CopyEndDevice(tc.Device)
-			fp := FrequencyPlan(dev.FrequencyPlanID)
-			err := adaptDataRate(
-				log.NewContext(test.ContextWithTB(test.Context(), t), test.GetLogger(t)),
-				dev,
-				LoRaWANBands[fp.BandID][dev.LoRaWANPHYVersion],
-				ttnpb.MACSettings{},
-			)
-			if !a.So(err, should.Equal, tc.Error) {
-				t.Fatalf("ADR failed with: %s", err)
-			}
-			expected := CopyEndDevice(tc.Device)
-			if tc.DeviceDiff != nil {
-				tc.DeviceDiff(expected)
-			}
-			a.So(dev, should.Resemble, expected)
+		test.RunSubtest(t, test.SubtestConfig{
+			Name:     tc.Name,
+			Parallel: true,
+			Func: func(ctx context.Context, t *testing.T, a *assertions.Assertion) {
+				dev := CopyEndDevice(tc.Device)
+				fp := FrequencyPlan(dev.FrequencyPlanID)
+				err := adaptDataRate(ctx, dev, LoRaWANBands[fp.BandID][dev.LoRaWANPHYVersion], ttnpb.MACSettings{})
+				if !a.So(err, should.Equal, tc.Error) {
+					t.Fatalf("ADR failed with: %s", err)
+				}
+				expected := CopyEndDevice(tc.Device)
+				if tc.DeviceDiff != nil {
+					tc.DeviceDiff(expected)
```

Comment:
Note the race condition here. `tc` must be copied before `test.RunSubtest` is run for all parallel test cases. That is done in https://github.com/TheThingsNetwork/lorawan-stack/pull/3052 and very hard to cherry-pick. Meanwhile I disabled parallel execution in this PR.

Reply:
~You've mentioned that you have disabled parallel execution, but I still see `Parallel: true` on line 476.  Is this intended ?~ See the diff bellow.

I believe you can add `tc := tc` from #3052 and it won't cause an issue during merge/rebase (you will still have to `ok` the conflict tho, but at the same time you will have a diff from the disabled parallel execution).

Reply:
This has to be done everywhere and for some test cases it's not just `tc := tc` that's necessary. This requires manual process, which is quite time-consuming, even with editor macros. The files are too different to simply checkout at this point and search for necessary diffs/generating appropriate diffs with `--patch` is a waste of time